### ISSUE TITLE
[BUG] Write blob data to blob storage fails randomly

### ISSUE THREAD

**Issue by omerbilgin** [Azure/azure-sdk-for-java#8079]
**Describe the bug**
While writing data to blob storage the library fails randomly with a nullpointerexception or indexarrayoutofbounds exception (depending on the version of azure.core). The data is written to blob storage so it seems that it is failing after completion of writing.

***Exception or Stack Trace***
**azure-core: 1.1.0**

> java.lang.NullPointerException, 	at java.base/java.util.ServiceLoader$ModuleServicesLookupIterator.hasNext(ServiceLoader.java:1067), 	at java.base/java.util.ServiceLoader$2.hasNext(ServiceLoader.java:1299), 	at java.base/java.util.ServiceLoader$3.hasNext(ServiceLoader.java:1384), 	at com.azure.core.util.tracing.TracerProxy.setSpanName(TracerProxy.java:77), 	at com.azure.core.http.rest.RestProxy.startTracingSpan(RestProxy.java:200), 	at com.azure.core.http.rest.RestProxy.invoke(RestProxy.java:134), 	at com.azure.storage.blob.implementation.$Proxy35.upload(Unknown Source), 	at com.azure.storage.blob.implementation.BlockBlobsImpl.uploadWithRestResponseAsync(BlockBlobsImpl.java:203), 	at com.azure.storage.blob.specialized.BlockBlobAsyncClient.uploadWithResponse(BlockBlobAsyncClient.java:201), 	at com.azure.storage.blob.specialized.BlockBlobClient.uploadWithResponse(BlockBlobClient.java:232),

**azure-core: 1.2.0**

> Exception in thread "20-1418141131" java.lang.ArrayIndexOutOfBoundsException: Index 2 out of bounds for length 2, 	at java.base/java.lang.CompoundEnumeration.next(ClassLoader.java:3022), 	at java.base/java.lang.CompoundEnumeration.hasMoreElements(ClassLoader.java:3031), 	at java.base/java.util.ServiceLoader$LazyClassPathLookupIterator.nextProviderClass(ServiceLoader.java:1202), 	at java.base/java.util.ServiceLoader$LazyClassPathLookupIterator.hasNextService(ServiceLoader.java:1220), 	at java.base/java.util.ServiceLoader$LazyClassPathLookupIterator.hasNext(ServiceLoader.java:1264), 	at java.base/java.util.ServiceLoader$2.hasNext(ServiceLoader.java:1299), 	at java.base/java.util.ServiceLoader$3.hasNext(ServiceLoader.java:1384), 	at com.azure.core.util.tracing.TracerProxy.start(TracerProxy.java:37), 	at com.azure.core.http.rest.RestProxy.startTracingSpan(RestProxy.java:201), 	at com.azure.core.http.rest.RestProxy.invoke(RestProxy.java:134), 	at com.azure.storage.blob.implementation.$Proxy35.upload(Unknown Source), 	at com.azure.storage.blob.implementation.BlockBlobsImpl.uploadWithRestResponseAsync(BlockBlobsImpl.java:203), 	at com.azure.storage.blob.specialized.BlockBlobAsyncClient.uploadWithResponse(BlockBlobAsyncClient.java:201), 	at com.azure.storage.blob.specialized.BlockBlobClient.uploadWithResponse(BlockBlobClient.java:209), 	at com.tomtom.autostream.persistence.azure.AzureBlobWriteDataResource.write(AzureBlobWriteDataResource.java:61), 	at com.tomtom.autostream.toolkit.tiledata.DataResourceTileDataWriter.write(DataResourceTileDataWriter.java:44), 	at com.tomtom.autostream.toolkit.datamodel.LayerProducer.storeTileData(LayerProducer.java:154), 	at com.tomtom.autostream.toolkit.datamodel.LayerProducer.createOutput(LayerProducer.java:131), 	at com.tomtom.autostream.toolkit.datamodel.LayerProducer.generateTilePerLayerFormat(LayerProducer.java:121), 	at com.tomtom.autostream.toolkit.datamodel.LayerProducer.generateBaseLevelTile(LayerProducer.java:78), 	at com.tomtom.autostream.toolkit.datamodel.LayerProducer.generateTilePerLayerFormat(LayerProducer.java:42), 	at com.tomtom.autostream.toolkit.datamodel.TileDataProducer.lambda$produceTilePerLayer$1(TileDataProducer.java:72), 	at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:195), 	at com.google.common.collect.CollectSpliterators$1.lambda$forEachRemaining$1(CollectSpliterators.java:120), 	at com.google.common.collect.CollectSpliterators$1.lambda$forEachRemaining$1(CollectSpliterators.java:120), 	at java.base/java.util.TreeMap$EntrySpliterator.forEachRemaining(TreeMap.java:2962), 	at com.google.common.collect.CollectSpliterators$1.forEachRemaining(CollectSpliterators.java:120), 	at com.google.common.collect.CollectSpliterators$FlatMapSpliterator.lambda$forEachRemaining$1(CollectSpliterators.java:361), 	at java.base/java.util.TreeMap$EntrySpliterator.forEachRemaining(TreeMap.java:2962), 	at com.google.common.collect.CollectSpliterators$FlatMapSpliterator.forEachRemaining(CollectSpliterators.java:357), 	at com.google.common.collect.CollectSpliterators$1.forEachRemaining(CollectSpliterators.java:120), 	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:484), 	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:474), 	at java.base/java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:913), 	at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234), 	at java.base/java.util.stream.ReferencePipeline.collect(ReferencePipeline.java:578), 	at com.tomtom.autostream.toolkit.datamodel.TileDataProducer.produceTilePerLayer(TileDataProducer.java:73), 	at com.tomtom.autostream.toolkit.datamodel.TileDataProducer.produce(TileDataProducer.java:40), 	at com.tomtom.autostream.toolkit.compiler.TileCompiler.lambda$compileTask$1(TileCompiler.java:153), 	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128), 	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628), 	at java.base/java.lang.Thread.run(Thread.java:834)]] 

**To Reproduce**
run the code snippet several times. at least third or fourth try, it will throw the exception.

***Code Snippet***
```
public void write(URI resourceUri, byte[] bytes) {
    var bucket = bucket(resourceUri);
    var key = key(resourceUri);

    validateUrlInBucket(bucket, key);

    var encodedBytes = compress ? GzipEncoding.compress(bytes) : bytes;
    var blobClient = getBlockBlobClient(bucket, key);

    try (var stream = new ByteArrayInputStream(encodedBytes)) {
      var blobRequestConditions = new BlobRequestConditions();
      // If key exists in container this will cause an exception.
      blobRequestConditions.setIfNoneMatch("*");
      blobClient.uploadWithResponse(
          stream,
          encodedBytes.length,
          createHttpHeaders(),
          null,
          null,
          null,
          blobRequestConditions,
          null,
          null);
    } catch (IOException | BlobStorageException e) {
      throw new WriteResourceException(
          String.format(
              "Storing data of length: %s bytes to Azure Blob Storage failed, tileKey: %s",
              bytes.length, key),
          e);
    }
}
```

**Expected behavior**
writing to azure blob storage must finish successfully without throwing an exception

**Screenshots**

**Setup (please complete the following information):**
 - OS:Ubuntu 18.04 LTS
 - IDE: IntelliJ
 - azure-sdk: 1.30.0
-  azure-blob-storage: 12.3.0
-  azure-core: 1.2.0 / 1.1.0

**Additional context**

**Comment by rickle-msft** [comment]
@alzimmermsft It looks like the stack trace is pointing to some problems with dynamically loading some class related to tracing? Can you take a look, please?

**Comment by joshfree** [comment]
Thank you for reporting this reliability issue @omerbilgin.  This appears to be related to OpenTelemetry tracing support (https://github.com/Azure/azure-sdk-for-java/blob/master/sdk/core/azure-core-tracing-opentelemetry/README.md); @alzimmermsft can you please investigate and update the thread with your findings?

/cc @samvaity @lmolkova as FYI

**Comment by samvaity** [comment]
@omerbilgin 

Thank you for reporting this issue.
To reproduce the issue on our end and for the most reliable resolution approach it would help if you could provide the following:

1. A sample test app for reproducing the above issue.
2. Could you also provide your pom.xml configuration?
3. Was there any attempt to enable tracing?

**Comment by carniwar** [comment]
Hi @samvaity 

Sorry, but we could not reproduce it in an app without our code and we're not allowed to post company code in Github. But here is some better explanation about our scenario.

This problem is during a call to `com.azure.storage.blob.specialized.BlockBlobClient.uploadWithResponse`. Azure SDK uses a RestProxy wrapping the HTTP call and check for Tracers to start tracing on it. In TracerProxy, there is a static property to the list of Tracers in the project. But list is retrieved via ServiceLoader in a lazy way. According to the documentation of this class (https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/util/ServiceLoader.html) it's not safe to use it in multi thread environments. This means that this call, and all others using RestProxy), are not thread safe.

Is there a thread safe way of using it? We need to parallelize this upload so it doesn't turn into a bottle neck in our software.

A possible solution for this is on the PR: https://github.com/Azure/azure-sdk-for-java/pull/8442. We checked and it works for us.

**Comment by alzimmermsft** [comment]
This should be fixed by PR #8446, this was released with Azure Core 1.3.0.
